addons/hr_work_entry_holidays/models/hr_contract.py

```diff
@@ -46,30 +47,32 @@ def _get_interval_leave_work_entry_type(self, interval, leaves, bypassing_codes)
         # global time off first (eg: Public Holiday > Home Working)
         self.ensure_one()
         if 'work_entry_type_id' in interval[2] and interval[2].work_entry_type_id.code in bypassing_codes:
-            return interval[2].work_entry_type_id
+            return [(interval[0], interval[1], interval[2].work_entry_type_id)]
 
         interval_start = interval[0].astimezone(pytz.utc).replace(tzinfo=None)
         interval_stop = interval[1].astimezone(pytz.utc).replace(tzinfo=None)
-        including_rcleaves = [l[2] for l in leaves if l[2] and interval_start >= l[2].date_from and interval_stop <= l[2].date_to]
-        including_global_rcleaves = [l for l in including_rcleaves if not l.holiday_id]
-        including_holiday_rcleaves = [l for l in including_rcleaves if l.holiday_id]
-        rc_leave = False
-
-        # Example: In CP200: Long term sick > Public Holidays (which is global)
-        if bypassing_codes:
-            bypassing_rc_leave = [l for l in including_holiday_rcleaves if l.holiday_id.holiday_status_id.work_entry_type_id.code in bypassing_codes]
-        else:
-            bypassing_rc_leave = []
-
-        if bypassing_rc_leave:
-            rc_leave = bypassing_rc_leave[0]
-        elif including_global_rcleaves:
-            rc_leave = including_global_rcleaves[0]
-        elif including_holiday_rcleaves:
-            rc_leave = including_holiday_rcleaves[0]
-        if rc_leave:
-            return self._get_leave_work_entry_type_dates(rc_leave, interval_start, interval_stop, self.employee_id)
-        return self.env.ref('hr_work_entry_contract.work_entry_type_leave')
+        work_entries_to_create = []
+        leaves_over_interval = [
+            l for l in leaves if l[2]
+            and (interval[0] <= l[0] and interval[1] >= l[1])
+            or (interval[0] >= l[0] and interval[1] <= l[1])
+        ]
+
+        # Priority rules if leaves overlap (Example: In CP200: Long term sick (which is bypassing) > Public Holidays (which is global))
+        bypassing_leaves = WorkIntervals([l for l in leaves_over_interval if self._get_leave_work_entry_type(l[2]).code in bypassing_codes])
+        global_leaves = WorkIntervals([l for l in leaves_over_interval if not l[2].holiday_id]) - bypassing_leaves
+        other_leaves = WorkIntervals(leaves_over_interval) - bypassing_leaves - global_leaves
+
+        for leave in bypassing_leaves | global_leaves | other_leaves:
+            if leave[1] <= interval[0] or leave[0] >= interval[1]:
+                continue
+            if leave[2]:
+                entry_type = self._get_leave_work_entry_type_dates(leave[2][0], interval_start, interval_stop, self.employee_id)
+            else:
+                entry_type = self.env.ref('hr_work_entry_contract.work_entry_type_leave')
+            work_entries_to_create.append((max(leave[0], interval[0]), min(leave[1], interval[1]), entry_type))
+
+        return work_entries_to_create
```

Comment:
You can't really change the return type of a method like this.

Other options are:
- factor out the logic into a new method, which may get called by this one
- modify the call to it so that you can ensure there's only a single `work_entry_type_id` to return for the interval given
- allow returning multiple `work_entry_type_id`s, and handle them accordingly afterwards

Reply:
Thanks for the feedback.

The first alternative maybe could have worked but I think that would take a lot of superfluous computation to make it fit a faulty mold. at this point it's easier to just use a new function imo. Also I propagated the change to the overrides, I'll need to rework the hk localization one as well because it seems to suffer from the same problem.

I left in the old function for the sake of not wrecking customers' customization but it is only workable if we assume that the minimum unit of time for leaves is the same as that of our work entries (Which is typically not the case unless you configure your working schedule and every leave type to fit the criterion). I think there is an argument to be made for removing it in master.